pkg/abcupdater/metrics.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2024 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package abcupdater
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"regexp"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/sethvargo/go-envconfig"
+
+	"github.com/abcxyz/abc-updater/pkg/abcupdater/localstore"
+	"github.com/abcxyz/pkg/logging"
+)
+
+const (
+	installIDFileName = "id.json"
+)
+
+var regExUUID = regexp.MustCompile("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")
+
+type metricsConfig struct {
+	ServerURL string `env:"ABC_UPDATER_METRICS_URL,default=https://abc-updater-metrics.tycho.joonix.net"`
+}
+
+// MetricsInfo are the parameters for sending metrics.
+type MetricsInfo struct {
+	// The ID of the application to check.
+	AppID string
+
+	// The version of the app to check for updates.
+	// Should be of form vMAJOR[.MINOR[.PATCH[-PRERELEASE][+BUILD]]] (e.g., v1.0.1)
+	Version string
+
+	// An optional Lookuper to load envconfig structs. Will default to os environment variables.
+	Lookuper envconfig.Lookuper
+
+	Metrics map[string]int
```

Comment:
Should this be a private field?

Reply:
I don't think so, this struct is provided by consumer of the library, and field needs to be able to be set by them. We could add a bunch of methods to manipulate, but seems overcomplicated.

Reply:
Why wouldn't we just have `WriteMetric(string, int)`? I'm confused why we'd let the client mutate this structure (especially when it pushes locking and concurrency onto their plate).

Reply:
Hmm, we could just make it a single metric, or do something like `WriteMetrics(string ...)`

@drevell Any opinions, you are the most likely first customer

Reply:
`WriteMetric(string,int)` seems fine to me. I don't see that it solves any concurrency problems though, unless we also add locking.

Reply:
OK, I'll do that then